src/murmur/Messages.cpp

```diff
@@ -77,6 +77,65 @@
 		sendMessage(uSource, mppd); \
 	}
 
+/// A helper class for managing temporary access tokens
+/// It will add the tokens in the comstructor and remove them again in the destructor effectively
+/// turning the tokens into a scope-based property
+class TemporaryAccessTokenHelper {
+	protected:
+		ServerUser *affectedUser;
+		QStringList qslTemporaryTokens;	
+		Server *server;
+
+	public:
+		TemporaryAccessTokenHelper(ServerUser *affectedUser, const QStringList &tokens, Server *server) : affectedUser(affectedUser),
+			qslTemporaryTokens(tokens), server(server) {
+			// Add the temporary tokens
+			QMutableStringListIterator it(this->qslTemporaryTokens);
+
+			{
+				QMutexLocker qml(&server->qmCache);
+
+				while (it.hasNext()) {
+					QString &token = it.next();
+
+					// If tokens are treated case-insensitively, transform all temp. tokens to lowercase first
+					if (Group::accessTokenCaseSensitivity == Qt::CaseInsensitive) {
+						token = token.toLower();
+					}
+
+					if (!this->affectedUser->qslAccessTokens.contains(token, Group::accessTokenCaseSensitivity)) {
+						// Add token
+						this->affectedUser->qslAccessTokens << token;
+					} else {
+						// It appears, as if the user already has this token set -> it's not a temporary one or a duplicate
+						it.remove();
+					}
+				}
+			}
+
+			if (!this->qslTemporaryTokens.isEmpty()) {
+				// Clear the cache in order for tokens to take effect
+				server->clearACLCache(this->affectedUser);
+			}
+		}
+
+		~TemporaryAccessTokenHelper() {
+			if (!this->qslTemporaryTokens.isEmpty()) {
+				{
+					QMutexLocker qml(&server->qmCache);
+
+					// remove the temporary tokens
+					foreach(QString token, this->qslTemporaryTokens) {
+						this->affectedUser->qslAccessTokens.removeOne(token);
```

Comment:
I would use `QStringList::removeAll()`.

Reply:
In principle there shouldn't be duplicates to begin with. RemoveAll will always have to iterate all elements in the list whereas removeOne can stop once it has found the respective element.